src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl_ModelComputation.cs

```diff
@@ -0,0 +1,257 @@
+        /// </summary>
+        private async ValueTask<DocumentSymbolModel?> UpdateUIAsync(ImmutableSegmentedList<bool> unused, CancellationToken cancellationToken)
+        {
+            var model = await _computeUIModelQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
+            if (model is null)
+                return null;
+
+            // Switch to the UI thread to get the current search query and latest active text view.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var searchQuery = searchBox.Text;
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return null;
+
+            // Switch to the threadpool to filter and sort the DocumentSymbolItems.
+            await TaskScheduler.Default;
+
+            var updatedDocumentSymbolItems = model.DocumentSymbolItems;
+
+            if (!string.IsNullOrWhiteSpace(searchQuery))
+                updatedDocumentSymbolItems = DocumentOutlineHelper.Search(updatedDocumentSymbolItems, searchQuery, cancellationToken);
+
+            updatedDocumentSymbolItems = DocumentOutlineHelper.Sort(updatedDocumentSymbolItems, SortOption, cancellationToken);
+
+            StartHightlightNodeTask(ExpansionOption.NoChange);
+
+            return new DocumentSymbolModel(updatedDocumentSymbolItems, model.OriginalSnapshot);
+        }
+
+        /// <summary>
+        /// Starts a new task to highlight the symbol node corresponding to the current caret position in the editor, expand/collapse
+        /// nodes (if applicable), and updates the UI.
+        /// </summary>
+        private void StartHightlightNodeTask(ExpansionOption expansionOption)
+        {
+            _highlightAndExpandNodesQueue.AddWork(expansionOption);
+        }
+
+        /// <summary>
+        /// Highlights the symbol node corresponding to the current caret position in the editor, expands/collapses nodes, then updates the UI.
+        /// </summary>
+        private async ValueTask HightlightNodeAsync(ImmutableSegmentedList<ExpansionOption> expansionOption, CancellationToken cancellationToken)
+        {
+            if (expansionOption.IsDefault || expansionOption.IsEmpty)
+                return;
+
+            var model = await _updateUIModelQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
+            if (model is null)
+                return;
+
+            // Switch to the UI thread to get the current caret point and latest active text view.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return;
+
+            var caretPoint = activeTextView.GetCaretPoint(activeTextView.TextBuffer);
+            if (!caretPoint.HasValue)
+                return;
```

Comment:
i actually don't know what happens here, but if this text view has lost focus (say because two views are visible), and the user switches to the other, does GetCaretPoint fail in the first view?

Reply:
It'll still work.